### PR TITLE
Feature:  Support Apigee as a deployment target for APIs

### PR DESCRIPTION
**Pull request by pspani**
### Feature: 

Support [Apigee](https://apigee.com/api-management/#/homepage) as a deployment target for APIs.

### Description of the Change

Support [Apigee](https://apigee.com/api-management/#/homepage) as a deployment target for APIs.

### Benefits

- Jazz admins will now be able to offer more than one deployment targets for deploying their API Gateway endpoints
- Developers can now choose Apigee as their API deployment target without worrying about the underlying Apigee infrastructure, setup & management.
- Existing AWS API Gateway service will still be supported as one of the API Gateway deployment targets.

### Possible Drawbacks

NA

### Applicable Issues

NA

### CODE REVIEW

File: misc/wiki/swagger/swagger.json [hunk 1]
```diff
@@ -2,17 +2,18 @@
   "swagger": "2.0",
   "info": {
     "version": "2017-01-30T23:30:18Z",
-    "title": "{api_deployment_node_title}",
+    "title": "{apiDeploymentNodeTitle}",
```
Comment (supritat) [inline misc/wiki/swagger/swagger.json]: why this got changed? This should be same for core and user created service right? Are we using this in any place?

File: builds/jazz-build-module/service-configuration-data-loader.groovy [hunk 2]
```diff
@@ -440,77 +492,73 @@ def loadServiceConfigurationData() {
...
+                    sh "sed -i -- 's/{bb_username}/${UNAME}/g' ./config/stg-config.json"
+                    sh "sed -i -- 's/{bb_username}/${UNAME}/g' ./config/prod-config.json"
+
+                    sh "sed -i -- 's/{bb_password}/${PWD}/g' ./config/dev-config.json"
+                    sh "sed -i -- 's/{bb_password}/${PWD}/g' ./config/stg-config.json"
+                    sh "sed -i -- 's/{bb_password}/${PWD}/g' ./config/prod-config.json"
+                }
+            }
 
-        if ((config_loader.SLACK.ENABLE_SLACK == "true") && (service_name.trim() == "jazz_slack-channel")) {
```
Comment (devsatishm) [inline builds/jazz-build-module/service-configuration-data-loader.groovy]: We'll need these changes.

Overall, I believe something went wrong during the merge of this file here. Please take a look at the diff again and add code segments that are removed during the branch sync.

File: builds/jazz-build-module/service-configuration-data-loader.groovy [hunk 3]
```diff
@@ -440,77 +492,73 @@ def loadServiceConfigurationData() {
...
-            sh "sed -i -- 's/{svc_acc_id}/${config_loader.SLACK.SLACK_SVC_ID}/g' ./config/dev-config.json"
-            sh "sed -i -- 's/{svc_acc_id}/${config_loader.SLACK.SLACK_SVC_ID}/g' ./config/stg-config.json"
-            sh "sed -i -- 's/{svc_acc_id}/${config_loader.SLACK.SLACK_SVC_ID}/g' ./config/prod-config.json"
-
         }
-    } catch (e) {
-        echo "error occured while loading service configuration: " + e.getMessage()
-        error "error occured while loading service configuration: " + e.getMessage()
-    }
+	} catch (e) {
```
Comment (devsatishm) [inline builds/jazz-build-module/service-configuration-data-loader.groovy]: Fix formatting.

File: builds/jazz-build-module/service-configuration-data-loader.groovy [hunk 4]
```diff
@@ -440,77 +492,73 @@ def loadServiceConfigurationData() {
...
+		echo "error occured while loading service configuration: " + e.getMessage()
+		error "error occured while loading service configuration: " + e.getMessage()
+	}
 }
 
 def setRoleARN(roleArn){
-    role_arn = roleArn
+	role_arn = roleArn
```
Comment (devsatishm) [inline builds/jazz-build-module/service-configuration-data-loader.groovy]: Fix formatting (here & below).

File: builds/jazz-build-module/service-configuration-data-loader.groovy [hunk 5]
```diff
@@ -537,18 +585,18 @@ def setEventSourceMapping(eventSourceArn, config) {
...
-            def rd = sh(script: "openssl rand -hex 4", returnStdout: true).trim()
-            sh "aws lambda add-permission --function-name arn:aws:lambda:${region}:${role_id}:function:${function_name} --statement-id lambdaFxnPermission${rd} --action lambda:* --principal logs.${region}.amazonaws.com --region ${region}"
-            echo "set permission for cloud-logs-streamer - success"
-        } catch (ss) {
-            //ignore if already registered permissions
-            echo(ss)
-            echo "set permission for cloud-logs-streamer - ok np"
-        }
-    }
+	if (config['service'] == "cloud-logs-streamer") {
```
Comment (devsatishm) [inline builds/jazz-build-module/service-configuration-data-loader.groovy]: Revert the changes, I think tab was used to format this function while rest of the code used spaces. Let's use one convention (spaces) across the board.

File: builds/jenkins-build-pack-api/Jenkinsfile [hunk 6]
```diff
@@ -210,11 +212,13 @@ node() {
 			events.sendStartedEvent('UPDATE_ENVIRONMENT', "Environment status update event for ${env_key} deployment", environmentDeploymentMetadata.generateEnvironmentMap("deployment_started", environment_logical_id, null), environment_logical_id)
 
 			def endpointUrl = null
+			def api_deployment_target
 			def aws_api_id = null
 
 			if (env_metadata["AWS_API_ID"] != null) {
 				aws_api_id = env_metadata["AWS_API_ID"]
 			}
+      def swaggerDocUrl=""
```
Comment (devsatishm) [inline builds/jenkins-build-pack-api/Jenkinsfile]: Fix formatting.

File: builds/jenkins-build-pack-api/Jenkinsfile [hunk 7]
```diff
@@ -245,75 +249,110 @@ node() {
...
+								break;
+
+							case 'gcp_apigee':
+								events.sendStartedEvent('DEPLOY_TO_GCP_APIGEE', 'Deployment started for APIGEE', context_map, environment_logical_id)
+								apiHostName ="${configLoader.APIGEE.API_ENDPOINTS[env_key].SERVICE_HOSTNAME}"
+								updateSwagger(apiHostName, environment_logical_id, current_environment, env_key, config)
+								endpointUrl = apigeeModule.deploy("swagger/swagger.json", lambdaARN, env_key, environment_logical_id, config, context_map)
+								echo "API Proxy created at: ${endpointUrl}"
+								events.sendCompletedEvent('DEPLOY_TO_GCP_APIGEE', 'Deployment complete for APIGEE', context_map, environment_logical_id)
+                //adding svc, domain and environment as the provider id since that is used as apiproxy(resource)
```
Comment (devsatishm) [inline builds/jenkins-build-pack-api/Jenkinsfile]: Fix formatting.

File: builds/delete-serverless-service-build-pack/Jenkinsfile [hunk 8]
```diff
@@ -1209,6 +1222,12 @@ def loadBuildModules(buildModuleUrl){
 		sonarModule = load "sonar-module.groovy"
 		echo "Sonar module loaded successfully."
 
+		if (configLoader.APIGEE && configLoader.APIGEE.ENABLED) {
```
Comment (devsatishm) [inline builds/delete-serverless-service-build-pack/Jenkinsfile]: if(configLoader.APIGEE && configLoader.APIGEE.ENABLE_APIGEE instanceof Boolean && configLoader.APIGEE.ENABLE_APIGEE) {...

File: builds/jazz-build-module/service-configuration-data-loader.groovy (hunk 2 above)
Comment (pspani) [reply builds/jazz-build-module/service-configuration-data-loader.groovy]: these lines were removed as part of #535. I will be adding them back in

File: builds/jazz-build-module/service-configuration-data-loader.groovy [hunk 9]
```diff
@@ -41,32 +42,51 @@ def initialize(config, role_arn, region, role_id, jenkins_url, current_environme
...
+        def apiKeyProd = utilModule.getAPIIdForCore(config_loader.AWS.API["PROD"])
+
         if (fileExists('swagger/swagger.json')) {
             //Swagger SEDs
             echo "Updating the Swagger SEDs"
             sh "sed -i -- 's/{conf-region}/${region}/g' ./swagger/swagger.json"
             sh "sed -i -- 's/{conf-accId}/${role_id}/g' ./swagger/swagger.json"
         }
 
-        if ((config_loader.SLACK.ENABLE_SLACK == "true") && (service_name.trim() == "jazz_is-slack-channel-available")) {
```
[the remark this answers is not in the document]
Comment (pspani) [reply builds/jazz-build-module/service-configuration-data-loader.groovy]: agreed... putting them back in

File: core/jazz_create-serverless-service/swagger/swagger.json [hunk 10]
```diff
@@ -309,9 +309,27 @@
                     "type": "string",
                     "description":"Slack channel to receive all the service notifications"
                 },
-				"require_internal_access": {
+				        "require_internal_access": {
```
Comment (devsatishm) [inline core/jazz_create-serverless-service/swagger/swagger.json]: Fix formatting.

File: core/jazz_services/config/global-config.json [hunk 11]
```diff
@@ -97,6 +97,13 @@
             "editable": true,
             "isEmptyAllowed": false,
             "type": "Object"
+        },
+        {
+            "key": "deployment_targets",
+            "required": false,
```
[the remark this answers is not in the document]
Comment (devsatishm) [reply core/jazz_services/config/global-config.json]: Yes, it should be true.

File: core/jazz_services/config/dev-config.json [hunk 12]
```diff
@@ -3,8 +3,8 @@
     "service_required_fields": ["service", "domain", "type", "created_by", "runtime", "status"],
     "service_update_fields": ["description", "email", "slack_channel", "tags", "repository", "status","metadata","is_public_endpoint"],
     "service_filter_params": ["service", "domain", "region", "type", "runtime", "created_by", "status"],
-    "service_optional_fields": ["description", "region", "repository", "email", "slack_channel", "tags", "status","metadata","is_public_endpoint"],
-    "service_return_fields": ["service", "domain", "type", "created_by", "runtime", "description", "region", "repository", "email", "slack_channel", "tags", "status","metadata","is_public_endpoint"],
+    "service_optional_fields": ["description", "region", "repository", "email", "slack_channel", "tags", "status","metadata","is_public_endpoint", "deployment_targets"],
```
Comment (devsatishm) [inline core/jazz_services/config/dev-config.json]: `deployment_targets` should be in `service_required_fields`. Please update other configurations & unit tests if required.

File: misc/wiki/swagger/swagger.json (hunk 1 above)
Comment (devsatishm) [reply misc/wiki/swagger/swagger.json]: We refer this file from https://github.com/tmobile/jazz/wiki/Supported-Service-Types-and-Code-Templates. These changes need to be reverted though.

Comment (devsatishm) [reply misc/wiki/swagger/swagger.json]: All the  changes except for adding `"operationId": "{operationId}"` should be reverted.

File: templates/api-template-python/swagger/swagger.json [hunk 13]
```diff
@@ -13,6 +13,7 @@
     "paths": {
         "/{service_name}": {
             "options": {
+                "operationId": "{operationId}",
```
Comment (devsatishm) [inline templates/api-template-python/swagger/swagger.json]: Can we do the same changes to swagger file in `go` template [here](https://github.com/tmobile/jazz/blob/develop/templates/api-template-go/swagger/swagger.json)?

Remove `x-amazon-apigateway-integration` section and add `operationId `.

File: builds/delete-serverless-service-build-pack/Jenkinsfile [hunk 14]
```diff
@@ -460,9 +473,9 @@ def updateSwaggerConfig() {
 			// TODO: the below couple of statements will be replaced with regular expression in very near future;
 			def roleId = roleARN.substring(roleARN.indexOf("::") + 2, roleARN.lastIndexOf(":"))
 
-			sh "sed -i -- 's/{conf-role}/${roleARN}/g' ./swagger/swagger.json"
-			sh "sed -i -- 's/{conf-region}/${region}/g' ./swagger/swagger.json"
-			sh "sed -i -- 's/{conf-accId}/${roleId}/g' ./swagger/swagger.json"
+			sh "sed -i -- 's/{confRole}/${roleARN}/g' ./swagger/swagger.json"
+			sh "sed -i -- 's/{confRegion}/${region}/g' ./swagger/swagger.json"
+			sh "sed -i -- 's/{confAccId}/${roleId}/g' ./swagger/swagger.json"
```
[the remark this answers is not in the document]
Comment (pspani) [reply builds/delete-serverless-service-build-pack/Jenkinsfile]: fixed

File: builds/jazz-build-module/service-configuration-data-loader.groovy (hunk 9 above)
[the remark this answers is not in the document]
Comment (pspani) [reply builds/jazz-build-module/service-configuration-data-loader.groovy]: fixed

File: builds/jazz-build-module/service-configuration-data-loader.groovy [hunk 15]
```diff
@@ -226,10 +245,10 @@ def loadServiceConfigurationData() {
             sh "sed -i -- 's/{jazz_admin_creds}/${config_loader.JAZZ.PASSWD}/g' ./config/prod-config.json"
         }
 
-        if ((config_loader.SLACK.ENABLE_SLACK == "true") && (service_name.trim() == "jazz_slack-event-handler")) {
```
[the remark this answers is not in the document]
Comment (pspani) [reply builds/jazz-build-module/service-configuration-data-loader.groovy]: fixed

File: builds/jazz-build-module/service-configuration-data-loader.groovy [hunk 16]
```diff
@@ -408,9 +463,6 @@ def loadServiceConfigurationData() {
             sh "sed -i -- 's/{region}/${region}/g' ./config/dev-config.json"
             sh "sed -i -- 's/{region}/${region}/g' ./config/stg-config.json"
             sh "sed -i -- 's/{region}/${region}/g' ./config/prod-config.json"
-        }
-
-        if (( service_name.trim() == "jazz_usermanagement") || (service_name.trim() == "jazz_admin")) {
```
[the remark this answers is not in the document]
Comment (pspani) [reply builds/jazz-build-module/service-configuration-data-loader.groovy]: agreed but partially. Repository base url moustache values are different for jazz_admin vs jazz_usermanagement. so implemented partially

File: builds/jazz-build-module/service-configuration-data-loader.groovy (hunk 2 above)
Comment (pspani) [reply builds/jazz-build-module/service-configuration-data-loader.groovy]: fixed

File: builds/jazz-build-module/service-configuration-data-loader.groovy (hunk 3 above)
Comment (pspani) [reply builds/jazz-build-module/service-configuration-data-loader.groovy]: fixed

File: builds/jazz-build-module/service-configuration-data-loader.groovy (hunk 4 above)
Comment (pspani) [reply builds/jazz-build-module/service-configuration-data-loader.groovy]: fixed

File: builds/jazz-build-module/service-configuration-data-loader.groovy (hunk 5 above)
Comment (pspani) [reply builds/jazz-build-module/service-configuration-data-loader.groovy]: fixed

File: builds/delete-serverless-service-build-pack/Jenkinsfile [hunk 17]
```diff
@@ -450,7 +463,7 @@ def writeServerlessFile(){
 def updateSwaggerConfig() {
 	try {
 		if (fileExists('swagger/swagger.json')) {
-			sh "sed -i -- 's/{service_name}/${service_config['service']}/g' swagger/swagger.json"
+			sh "sed -i -- 's/{serviceName}/${service_config['service']}/g' swagger/swagger.json"
```
[the remark this answers is not in the document]
Comment (pspani) [reply builds/delete-serverless-service-build-pack/Jenkinsfile]: fixed

File: builds/delete-serverless-service-build-pack/Jenkinsfile (hunk 14 above)
[the remark this answers is not in the document]
Comment (pspani) [reply builds/delete-serverless-service-build-pack/Jenkinsfile]: fixed

File: builds/delete-serverless-service-build-pack/Jenkinsfile (hunk 8 above)
Comment (pspani) [reply builds/delete-serverless-service-build-pack/Jenkinsfile]: fixed

File: core/jazz_services/config/dev-config.json (hunk 12 above)
Comment (pspani) [reply core/jazz_services/config/dev-config.json]: fixed

File: core/jazz_services/config/global-config.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (pspani) [reply core/jazz_services/config/global-config.json]: fixed

File: misc/wiki/swagger/swagger.json (hunk 1 above)
Comment (pspani) [reply misc/wiki/swagger/swagger.json]: fixed

File: builds/jenkins-build-pack-api/Jenkinsfile (hunk 6 above)
Comment (pspani) [reply builds/jenkins-build-pack-api/Jenkinsfile]: fixed

File: builds/jenkins-build-pack-api/Jenkinsfile (hunk 7 above)
Comment (pspani) [reply builds/jenkins-build-pack-api/Jenkinsfile]: fixed

File: core/jazz_create-serverless-service/swagger/swagger.json (hunk 10 above)
Comment (pspani) [reply core/jazz_create-serverless-service/swagger/swagger.json]: fixed

File: templates/api-template-python/swagger/swagger.json (hunk 13 above)
Comment (pspani) [reply templates/api-template-python/swagger/swagger.json]: fixed
